### PR TITLE
unwrap FormData in XMLHttpRequest.send if needed

### LINKED ISSUE #725: FormData wrapper breaks XMLHttpRequest#send(FormData) in Firefox
XMLHttpRequest does not unwrap the FormData object when sending, causing the following error in Firefox:

```
NS_ERROR_XPC_JSOBJECT_HAS_NO_FUNCTION_NAMED: JavaScript component does not have a method named: "available"'JavaScript component does not have a method named: "available"' when calling method: [nsIInputStream::available]
request.send(formData);
```

[Screenshot](https://cloud.githubusercontent.com/assets/613610/4053262/97a4ce10-2d77-11e4-9dea-17db0e6b8488.png)
<p>
Example:

```
<script src="platform.js"></script>
<script>
  function doIt() {
    var request = new XMLHttpRequest();
    request.open('POST', '//localhost:8000/handle');
    request.send(new FormData());
  }
</script>
<a href="javascript:doIt();">Click</a>
```

<p>
Removing the platform.js script tag or executing the following code fixes the issue:

```
var oldSend = XMLHttpRequest.prototype.send; XMLHttpRequest.prototype.send = function(arg) { oldSend.call(this, unwrap(arg)); };
```

Browser: Firefox 31.0, Mozilla Firefox for Ubuntu canonical - 1.0
Platform: 0.3.5
ShadowDOM: c4da63735ba6c00a7d6af5c1b118f84bd6a2e114

### PR DESCRIPTION
**Pull request by jmesserly**
fixes https://github.com/Polymer/polymer/issues/725

### CODE REVIEW

**Comment by jmesserly** [comment]
@arv mind taking a look?

Even though it's smallish, I created a few new files to match the existing style.

File: src/wrappers/XMLHttpRequest.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * license that can be found in the LICENSE file.
+ */
+
+(function(scope) {
+  'use strict';
+
+  var unwrapIfNeeded = scope.unwrapIfNeeded;
+  var XMLHttpRequest = window.XMLHttpRequest;
+
+  var originalSend = XMLHttpRequest && XMLHttpRequest.prototype.send;
```
Comment (arv) [inline src/wrappers/XMLHttpRequest.js]: You don't need to check if XMLHttpRequest is available. All browsers support it alreade.

``` js
  var unwrapIfNeeded = scope.unwrapIfNeeded;

  var originalSend = XMLHttpRequest.prototype.send;
```

File: src/wrappers/XMLHttpRequest.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  if (!originalSend) return;
```
Comment (arv) [inline src/wrappers/XMLHttpRequest.js]: remove this line

File: src/wrappers/XMLHttpRequest.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+  // Note: since we only need to adjust XHR.send, we just patch it instead of
```
Comment (arv) [inline src/wrappers/XMLHttpRequest.js]: Remove `Note:`

File: src/wrappers/XMLHttpRequest.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  // wrapping the entire object.
+  XMLHttpRequest.prototype.send = function(obj) {
+    // Unwrap FormData if needed
```
Comment (arv) [inline src/wrappers/XMLHttpRequest.js]: Useless comment

File: src/wrappers/XMLHttpRequest.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  'use strict';
+
+  var unwrapIfNeeded = scope.unwrapIfNeeded;
+  var XMLHttpRequest = window.XMLHttpRequest;
+
+  var originalSend = XMLHttpRequest && XMLHttpRequest.prototype.send;
+  if (!originalSend) return;
+
+  // Note: since we only need to adjust XHR.send, we just patch it instead of
+  // wrapping the entire object.
```
Comment (arv) [inline src/wrappers/XMLHttpRequest.js]: I wonder if there are more cases where we can skip the wrapper class? We should keep our eyes open.

**Comment by arv** [comment]
LGTM with nits

File: src/wrappers/XMLHttpRequest.js (hunk 1 above)
Comment (jmesserly) [inline src/wrappers/XMLHttpRequest.js]: Done.

File: src/wrappers/XMLHttpRequest.js (hunk 2 above)
Comment (jmesserly) [inline src/wrappers/XMLHttpRequest.js]: Done. seemed to go with previous comment :)

File: src/wrappers/XMLHttpRequest.js (hunk 3 above)
Comment (jmesserly) [inline src/wrappers/XMLHttpRequest.js]: Done

File: src/wrappers/XMLHttpRequest.js (hunk 4 above)
Comment (jmesserly) [inline src/wrappers/XMLHttpRequest.js]: So, I kind of liked this one. XHR.send takes a bunch of things most of which are not wrapped, so I can see someone looking at this code and thinking "why are we patching XHR?". Maybe I'll mention FormData in the comment above. My thinking was that we didn't think of the XHR<->FormData connection when patching FormData, so someone looking at this file might not grasp it immediately either.

**Comment by jmesserly** [comment]
PTAL

**Comment by arv** [comment]
Great

LGTM

**Comment by jmesserly** [comment]
awesome, thank you!

**Pull request merged**